src/language/customPathInstaller.ts

```diff
@@ -0,0 +1,152 @@
+import { promisify } from 'util';
+import * as os from 'os';
+import * as fs from 'fs';
+import { ExtensionContext, OutputChannel, window } from 'vscode';
+
+const existsAsync = promisify(fs.exists);
+import * as path from 'path';
+import { ConfigurationConstants, LanguageServerConstants } from '../constants';
+import Configuration from '../configuration';
+import { getPreferredVersion } from './dafnyInstallation';
+import { Executable } from 'vscode-languageclient/node';
+import { getDotnetExecutablePath } from '../dotnet';
+
+const mkdirAsync = promisify(fs.mkdir);
+
+export class CustomPathInstaller {
+  public constructor(
+    public readonly context: ExtensionContext,
+    public readonly statusOutput: OutputChannel
+  ) {}
+
+  // We cache the cli path so that we don't need to copy it every time.
+  private cliPath: string | undefined | null = null;
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  public async getExecutable(context: ExtensionContext, server: boolean, newArgs: string[], oldArgs: string[]): Promise<Executable | undefined> {
+    const configuredCliPath = await this.getCliPath(context);
+    if(!configuredCliPath) {
+      return undefined;
+    }
+
+    if(server) {
+      newArgs.unshift('server');
+    }
+    if(configuredCliPath.endsWith('.dll')) {
+      const { path: dotnetExecutable } = await getDotnetExecutablePath();
+      return {
+        command: dotnetExecutable,
+        args: [ configuredCliPath ].concat(newArgs)
+      };
+    } else {
+      return {
+        command: configuredCliPath,
+        args: newArgs
+      };
+    }
+  }
+
+  private async getCliPath(context: ExtensionContext): Promise<string> {
+    if(this.cliPath != null) {
+      return this.cliPath;
+    }
+    this.cliPath = await this.getCliPathUncached(context);
+    return this.cliPath!;
+  }
+
+  private async getCliPathUncached(context: ExtensionContext): Promise<string> {
+    let cliPathOverride = process.env['DAFNY_SERVER_OVERRIDE'] ?? '';
+    if(cliPathOverride) {
+      try {
+        cliPathOverride = await this.copyNecessaryDllsToTempFolder(cliPathOverride);
+      } catch(e: unknown) {
+        // Ignore if copying fails. We'll run the original.
+      }
+      window.showInformationMessage(`Using $DAFNY_SERVER_OVERRIDE = ${cliPathOverride} for the CLI path`);
+    }
+    if(cliPathOverride) {
+      return cliPathOverride;
+    }
+
+    const version = getPreferredVersion();
+    if(version !== LanguageServerConstants.Custom) {
+      return '';
```

Comment:
This method never returned an empty string before. Why is it doing that now?

Reply:
For strings it's safest to use an empty string as the "not set" value, since UI components will return '' when users clear the field. It can also return empty string on line 79 where it returns the UI value.